### PR TITLE
[cxx] wchar_t is not guint16, on Windows

### PR DESCRIPTION
**Pull request by jaykrell**
https://github.com/mono/mono/commit/675c94ec7a923a40c9e2dfdb098218f30226604b was incomplete.

https://jenkins.mono-project.com/job/x/12791/parsed_console/log.html

icall-windows.c: In function 'MonoStringHandle mono_icall_get_machine_name(MonoError*)':
icall-windows.c:65:75: error: invalid conversion from 'gunichar2* {aka wchar_t*}' to 'const guint16* {aka const short unsigned int*}' [-fpermissive]
   return mono_string_new_utf16_handle (mono_domain_get (), buf, len, error);
                                                                           ^
In file included from ../../mono/metadata/icall-internals.h:11:0,
                 from ../../mono/metadata/icall-windows-internals.h:13,
                 from icall-windows.c:14:
../../mono/metadata/object-internals.h:1963:1: note:   initializing argument 2 of 'MonoStringHandle mono_string_new_utf16_handle(MonoDomain*, const guint16*, gint32, MonoError*)'
 mono_string_new_utf16_handle (MonoDomain *domain, const guint16 *text, gint32 len, MonoError *error);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
icall-windows.c: In function 'MonoArray* mono_icall_get_environment_variable_names(MonoError*)':
icall-windows.c:136:101: error: invalid conversion from 'WCHAR* {aka wchar_t*}' to 'const guint16* {aka const short unsigned int*}' [-fpermissive]
     str = mono_string_new_utf16_checked (domain, env_string, (gint32)(equal_str - env_string), error);
                                                                                                     ^
In file included from ../../mono/metadata/icall-internals.h:11:0,
                 from ../../mono/metadata/icall-windows-internals.h:13,
                 from icall-windows.c:14:
../../mono/metadata/object-internals.h:1960:1: note:   initializing argument 2 of 'MonoString* mono_string_new_utf16_checked(MonoDomain*, const guint16*, gint32, MonoError*)'
 mono_string_new_utf16_checked (MonoDomain *domain, const guint16 *text, gint32 len, MonoError *error);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
icall-windows.c: In function 'void mono_icall_set_environment_variable(MonoString*, MonoString*)':
icall-windows.c:162:20: error: invalid conversion from 'mono_unichar2* {aka short unsigned int*}' to 'gunichar2* {aka wchar_t*}' [-fpermissive]
  utf16_name = name ? mono_string_chars (name) : NULL;
                    ^
icall-windows.c:168:34: error: invalid conversion from 'mono_unichar2* {aka short unsigned int*}' to 'gunichar2* {aka wchar_t*}' [-fpermissive]
  utf16_value = mono_string_chars (value);
                ~~~~~~~~~~~~~~~~~~^~~~~~~
icall-windows.c: In function 'MonoStringHandle mono_icall_get_windows_folder_path(int, MonoError*)':
icall-windows.c:188:76: error: invalid conversion from 'WCHAR* {aka wchar_t*}' to 'const guint16* {aka const short unsigned int*}' [-fpermissive]
   return mono_string_new_utf16_handle (mono_domain_get (), path, len, error);
                                                                            ^
In file included from ../../mono/metadata/icall-internals.h:11:0,
                 from ../../mono/metadata/icall-windows-internals.h:13,
                 from icall-windows.c:14:
../../mono/metadata/object-internals.h:1963:1: note:   initializing argument 2 of 'MonoStringHandle mono_string_new_utf16_handle(MonoDomain*, const guint16*, gint32, MonoError*)'
 mono_string_new_utf16_handle (MonoDomain *domain, const guint16 *text, gint32 len, MonoError *error);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

File: mono/metadata/object.c [hunk 1]
```diff
@@ -6415,7 +6415,7 @@ mono_string_empty_handle (MonoDomain *domain)
  * \returns A newly created string object which contains \p text.
  */
 MonoString *
-mono_string_new_utf16 (MonoDomain *domain, const guint16 *text, gint32 len)
+mono_string_new_utf16 (MonoDomain *domain, const gunichar2 *text, gint32 len)
```
Comment (lambdageek) [inline mono/metadata/object.c]: This is declared in object.h as
```c
MONO_API MONO_RT_EXTERNAL_ONLY
MonoString*
mono_string_new_utf16	    (MonoDomain *domain, const mono_unichar2 *text, int32_t len);
```

and `mono_unichar2` is guint16 (see `mono-publib.h`)

Comment (jaykrell) [reply mono/metadata/object.c]: Either way ok with me. 'g' is brief, `mono` matches, and must match.

This is all a breaking change, btw, however pretty rare, and the C++ compiler broke people the same way and more.

i.e.
```
MonoString (*function_pointer)(MonoDomain*, const guint16*, gint32 len) = mono_string_new_utf16;
```

must be changed to:
```MonoString (*function_pointer)(MonoDomain*, const gunichar2*, gint32 len) = mono_string_new_utf16;
```

The larger C++ breakage was the name mangling level.
The break is embedders, using C++, on Windows, taking address, w/o templatizing or casting.

Comment (jaykrell) [reply mono/metadata/object.c]: Similarly:

```
MONO_API mono_unichar2 *mono_string_chars  (MonoString *s);
```

```
const gunichar2 *p = mono_string_chars (s);
```

Where is the boundary that we transition to `g`?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build  OS X x64

**Comment by jaykrell** [comment]
Review?

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Pull request merged**